luminoth/models/ssd/ssd_target.py

```diff
@@ -0,0 +1,194 @@
+        cls_probs = probs[:, 1:]
+        max_cls_probs = tf.reduce_max(cls_probs, axis=1)
+
+        # Exclude boxes with IOU > `background_threshold_high` with any GT.
+        iou_less_than_bg_tresh_high_filter = tf.less_equal(
+            max_overlaps, self._background_threshold_high
+        )
+        bg_anchors = tf.less_equal(anchors_label, 0)
+        bg_overlaps_filter = tf.logical_and(
+            iou_less_than_bg_tresh_high_filter, bg_anchors
+        )
+
+        max_cls_probs = tf.where(
+            condition=bg_overlaps_filter,
+            x=max_cls_probs,
+            y=tf.fill(dims=anchors_label_shape, value=-1.),
+        )
+
+        # We calculate up to how many backgrounds we desire based on the
+        # final number of foregrounds and the hard minning ratio.
+        num_fg_mask = tf.greater(anchors_label, 0.0)
+        num_fg = tf.cast(tf.count_nonzero(num_fg_mask), tf.float32)
+
+        num_bg = tf.cast(num_fg * self._foreground_fraction, tf.int32)
+        top_k_bg = tf.nn.top_k(max_cls_probs, k=num_bg)
+
+        set_bg = tf.sparse_to_dense(
+            sparse_indices=top_k_bg.indices,
+            sparse_values=True, default_value=False,
+            output_shape=anchors_label_shape,
+            validate_indices=False
+        )
+
+        anchors_label = tf.where(
+            condition=set_bg,
+            x=tf.fill(dims=anchors_label_shape, value=0.),
+            y=anchors_label
+        )
+
+        # Next step is to calculate the proper bbox targets for the labeled
+        # anchors based on the values of the ground-truth boxes.
+        # We have to use only the anchors labeled >= 1, each matching with
+        # the proper gt_boxes
+
+        # Get the ids of the anchors that mater for bbox_target comparison.
+        is_anchor_with_target = tf.greater(
+            anchors_label, 0
+        )
+        anchors_with_target_idx = tf.where(
+            condition=is_anchor_with_target
+        )
+        # Get the corresponding ground truth box only for the anchors with
+        # target.
+        gt_boxes_idxs = tf.gather(
+            best_gtbox_for_anchors_idx,
+            anchors_with_target_idx
+        )
+        # Get the values of the ground truth boxes.
+        anchors_gt_boxes = tf.gather_nd(
+            gt_boxes[:, :4], gt_boxes_idxs
+        )
+        # We create the same array but with the anchors
+        anchors_with_target = tf.gather_nd(
+            all_anchors,
+            anchors_with_target_idx
+        )
+        # We create our targets with bbox_transform
+        bbox_targets = encode(
+            anchors_with_target,
+            anchors_gt_boxes,
+            variances=self._variances
+        )
+
+        # We unmap targets to anchor_labels (containing the length of
+        # anchors)
+        bbox_targets = tf.scatter_nd(
+            indices=anchors_with_target_idx,
+            updates=bbox_targets,
+            shape=tf.cast(tf.shape(all_anchors), tf.int64)
```

Comment:
tf.int or 32

Reply:
We use int64 because `indices` is int64, and this is because `indices` comes from an `tf.where` with no x and y arguments, which is hardcoded to return int64 by tensorflow.